apps/ios/SimpleX NSE/NotificationService.swift

```diff
@@ -254,23 +285,42 @@ class NotificationService: UNNotificationServiceExtension {
         deliverBestAttemptNtf()
     }
 
-    func addExpectedMessage(ntfMessage: UserNtfMessage) {
-        if let connEntity = ntfMessage.connEntity_,
-           let receiveEntityId = connEntity.id, ntfMessage.expectedMsg_ != nil {
-            let expectedMsgId = ntfMessage.expectedMsg_?.msgId
-            let receivedMsgId = ntfMessage.receivedMsg_?.msgId
-            logger.debug("NotificationService: addExpectedMessage: expectedMsgId = \(expectedMsgId ?? "nil", privacy: .private), receivedMsgId = \(receivedMsgId ?? "nil", privacy: .private)")
+    func addExpectedMessage(ntfConn: UserNtfConn) {
+        if let connEntity = ntfConn.connEntity_,
+           let receiveEntityId = connEntity.id, ntfConn.expectedMsg_ != nil {
+            let expectedMsgId = ntfConn.expectedMsg_?.msgId
+            logger.debug("NotificationService: addExpectedMessage: expectedMsgId = \(expectedMsgId ?? "nil", privacy: .private)")
             expectedMessages[receiveEntityId] = ExpectedMessage(
-                ntfMessage: ntfMessage,
+                ntfConn: ntfConn,
                 receiveConnId: connEntity.conn.agentConnId,
                 expectedMsgId: expectedMsgId,
                 allowedGetNextAttempts: 3,
-                msgBestAttemptNtf: ntfMessage.defaultBestAttemptNtf,
-                ready: false
+                msgBestAttemptNtf: ntfConn.defaultBestAttemptNtf,
+                ready: false,
+                shouldProcessNtf: false,
+                startedProcessingNewMsgs: false,
+                semaphore: DispatchSemaphore(value: 0)
             )
         }
     }
 
+    func checkCanStart(_ entityId: String) -> Bool {
+        return !NSEThreads.shared.activeThreads.contains(where: {
+            (tId, nse) in tId != threadId && nse.expectedMessages.contains(where: { $0.key == entityId })
+        })
+    }
+
+    func processDroppedNotifications(_ entityId: String) {
+        if !NSEThreads.shared.droppedNotifications.isEmpty {
+            let messagesToProcess = NSEThreads.shared.droppedNotifications.filter { (eId, _) in eId == entityId }
+            NSEThreads.shared.droppedNotifications.removeAll(where: { (eId, _) in eId == entityId })
+            for (index, (_, ntf)) in messagesToProcess.enumerated() {
+                if let t = threadId { logger.debug("NotificationService thread \(t, privacy: .private): entity \(entityId, privacy: .private): processing dropped notification \(index, privacy: .private)") }
+                processReceivedNtf(entityId, ntf, signalReady: false)
+            }
```

Comment:
```suggestion
            var i: Int = 0
            var toRemove: Set<Int> = []
            for i in 0..<NSEThreads.shared.droppedNotifications.count {
              let (eId, ntf) = NSEThreads.shared.droppedNotifications[I]
              if eId == entityId {
                if let t = threadId { logger.debug("NotificationService thread \(t, privacy: .private): entity \(entityId, privacy: .private): processing dropped notification \(i, privacy: .private)") }
                processReceivedNtf(entityId, ntf, signalReady: false)
                toRemove.insert(i)
              }
            }
            NSEThreads.shared.droppedNotifications.remove(at: toRemove)
```

Reply:
reverted this refactoring:
- there is no method remove(at: <set>)
- if we do "for i in <set> { remove(at: i) }" indexes will probably get mixed up
- remove(atOffsets: ...) uses IndexSet, not Set, seems like it's usually used for UI.. probably could use it but it's untested code